gstd/src/exec/async.rs

```diff
@@ -0,0 +1,69 @@
+// This file is part of Gear.
+
+// Copyright (C) Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{async_runtime, msg, MessageId};
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use gcore::exec;
+
+/// Delays message execution in asynchronous way for the specified number of
+/// blocks. It works pretty much like the `exec::wait_for` function, but allows
+/// to continue execution after the delay in the same handler. It is worth
+/// mentioning that the program state gets persisted inside the call, and the
+/// execution resumes with potentially different state.
+pub fn delay_for(block_count: u32) -> impl Future<Output = ()> {
+    MessageDelayFuture::new(msg::id(), exec::block_height().saturating_add(block_count))
+}
+
+struct MessageDelayFuture {
+    msg_id: MessageId,
+    till_block_number: u32,
+}
+
+impl MessageDelayFuture {
+    fn new(msg_id: MessageId, till_block_number: u32) -> Self {
+        Self {
+            msg_id,
+            till_block_number,
+        }
+    }
+}
+
+impl Future for MessageDelayFuture {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let current_block_number = exec::block_height();
+
+        async_runtime::locks().remove(self.msg_id, self.msg_id);
```

Comment:
what if we introduce for locks special lock id? 
I suppose that this will work incorrect for following case (please provide test for it):
*here goes some pseudo code*
```
let long = delay_for(10);
let short = delay_for(8);
join(long,short).await
```

await will wake us up in 8, isn't it or even end up with panic?
just because if we poll similar lock ids it will rewrite each other?